core/key.go

```diff
@@ -257,6 +264,11 @@ func NewCollectionSchemaVersionKey(schemaVersionId string) CollectionSchemaVersi
 	return CollectionSchemaVersionKey{SchemaVersionId: schemaVersionId}
 }
 
+func NewCollectionSchemaVersionKeyFromString(key string) CollectionSchemaVersionKey {
```

Comment:
I do not.  This is very simple, and tested by pretty much every integration test we have.  Unit tests would only slow down development (now and in the future).

Reply:
I had a closer look and I think that not unit testing our keys could cause bad changes to go unnoticed until the breaking change tests are executed which could be quite annoying.

Reply:
I'm not really in the mood to get back into the integration vs unit test conversation at the moment.  I have written a large amount on this topic before and you know my opinion on this matter.

Reply:
I don't think the mood of anyone is relevant to the conversation though. This is for the benefit of everyone on the team. And in this case a quick unit test would be a good approach.

Reply:
> And in this case a quick unit test would be a good approach.

You think that, and perhaps Islam thinks that.  But you also know that I really don't think that, and that I have had a lot of my time spent lately maintaining unit tests that have falsely failed when changing unrelated parts of the codebase.

Are you looking to re-open the conversation on unit vs integration tests here in this PR, or force me to make a change that you know I think will waste time?

Reply:
I think your point can be debated for tests that need to interact with multiple functions/types. In that case it can potentially be subjective. In this case however, it's a direct test of the key generation and its methods and it's the most direct and explicit way to flag if something is wrong with the key generation. Like I said, only the change detector would be warn us of something wrong but it would be quite vague as to what is actually wrong. Probably a `key does not exist` error or something of the sort.

Reply:
> In this case however, it's a direct test of the key generation and its methods and it's the most direct and explicit way to flag if something is wrong with the key generation.

A unit test of this function would not spot any failing in key generation, it would only re-describe in lengthier terms that this function does what the two lines in it say it does.

Reply:
There is more than one function here. The comment was on one line but it applies to all the added functions.

Reply:
> it would only re-describe in lengthier terms that this function does what the two lines in it say it does.

That is basically what every test we have does. Integration or otherwise. They describe in lengthier terms what functions and features do without needing a human to hand test everything.

We think testing these directly is relevant, quick, and the resulting test very easy to maintain. Not sure it's worth debating any further. Up to you.

Reply:
I understand Andy your frustration dealing with some unit tests, especially with those that I wrote with mocks for testing secondary indexes. I think it was a combination of me zealously trying to achieve high test coverage (I'm rethinking it now) and not ideal design that didn't let easily to get to certain code parts.

This code with keys on the other hand has no dependancies and therefore very straightforward to test with low maintenance in the future.

Reply:
> That is basically what every test we have does. Integration or otherwise. They describe in lengthier terms what functions and features do without needing a human to hand test everything.

Yes, the difference between unit and integration tests though is that integration tests test behaviour that is meaningful to end users, and unit tests do not.  

It really does not matter what this function does, so long as it behaves correctly given the parameters the rest of the code will provide it in production.  What that behaviour is does not matter.  Other conditions do not matter.

If this function changes, it will be because what it needs to do has changed.  In such a situation they unit tests would be useless, as they only describe what it used to do, not what it used to need to do, nor what it now needs to do.  Changing the behaviour of this function would essentially be a rip and replace, and they unit tests would essentially be ripped and replaced too.  The would be replaced with more useless unit tests that only describe what the production code does.  

They do not, and will not ever describe what it needs to do, because what it needs to do is totally imaginary and is not grounded in any real-world/user-visible use case.  They would be a totally pointless, vague, re-wording of the contents of the function, and would only get in the way if that imaginary use-case changes.

> I understand Andy your frustration dealing with some unit tests...

Whilst there is some overlap in my opinion of both this, and the index unit tests, the index unit tests that were removed were removed for different reasons to the ones I have just outlined in this discussion.  I'd prefer we don't confuse the two issues and will happily discuss the index tests in more details elsewhere (in this PR).

Reply:
We have two overall objective with testing:
1- Testing behaviour that is meaningful to the end user (like you mentioned above)
2- Reaching maximum code coverage so that all possible code paths have been tested (sometimes that is currently impossible to do just with integration tests). This is a goal that the team set for itself (significantly increase code coverage).

I'll end my input on this conversation with reminding you of the above objectives. If you believe that simply relying on integration tests in case contributes towards achieving the above goals, then I think we're good.

Reply:
>  If you believe that simply relying on integration tests in case contributes towards achieving the above objectives

I very strongly believe so.  Sometimes the integration tests will need to be more imaginative than our current core suite, for example chaos monkey like magic, or providing a deliberately faulty/test datastore implementation (the root datastore is something that can be publicly provided), but ultimately everything should be testable via public only access points.  If it isn't, it is dead code.

Reply:
just another note: it is practically impossible to reach a desired test coverage just with integration tests.
Just imagine setting up "a deliberately faulty/test datastore" that you want to trigger for a specific request at a specific code location. You saw it was hard to achieve it with unit tests, doing it with integration tests will be a nightmare.

And those tests helped me write better code and fix bugs (not reported) in the existing code. Like this:
```go
err := someAction()
if err != nil {
    return err
}
...
res :=rsultsIter.NextSync()
if res.Error != nil {
    return err // here is the problem
}
```
Usually we don't write test that such paths. But we can easily overlook the details and ship incorrect code.

Reply:
Such resilience tests don't need to be coded like traditional do-this, do-that tests. You can use this https://en.wikipedia.org/wiki/Chaos_engineering approach instead.  This can allow the tests to be somewhat proactive (searching for bugs) instead reactive (testing known pathways where bugs are expected).

Also a lot of our errors can be reached my manipulating the datastore values directly, no need to mess around with mocks.

Reply:
I also want to introduce a really magic testing layer, where randomized queries and setup data are fed in.  Traditionally you can do this using recorded real-world data/actions, but there has been significant movement in using 'AI' to generate stuff like queries for such things.